extensions/attributegen/plugin.cc

```diff
@@ -0,0 +1,233 @@
+  Status status =
+      JsonStringToMessage(configuration->toString(), &config, json_options);
+  if (status != Status::OK) {
+    LOG_WARN(absl::StrCat(
+        "Cannot parse 'attributegen' plugin configuration JSON string [YAML is "
+        "not supported]: ",
+        configuration->toString()));
+    incrementMetric(config_errors_, 1);
+    return false;
+  }
+
+  debug_ = config.debug();
+
+  cleanupAttributeGen();
+  auto init_status = initAttributeGen(config);
+  if (!init_status) {
+    incrementMetric(config_errors_, 1);
+    cleanupAttributeGen();
+  }
+
+  return init_status;
+}
+
+bool PluginRootContext::initAttributeGen(
+    const istio::attributegen::PluginConfig& config) {
+  for (const auto& attribute_gen_config : config.attributes()) {
+    EvalPhase phase = OnLog;
+    if (attribute_gen_config.phase() == istio::attributegen::ON_REQUEST) {
+      phase = OnRequest;
+    }
+    std::vector<Match> matches;
+
+    for (const auto& matchconfig : attribute_gen_config.match()) {
+      uint32_t token = 0;
+      if (matchconfig.condition().empty()) {
+        matches.push_back(Match("", 0, matchconfig.value()));
+        continue;
+      }
+      auto create_status = createExpression(matchconfig.condition(), &token);
+
+      if (create_status != WasmResult::Ok) {
+        LOG_WARN(absl::StrCat("Cannot create expression: <",
+                              matchconfig.condition(), "> for ",
+                              attribute_gen_config.output_attribute(),
+                              " result:", toString(create_status)));
+        return false;
+      }
+      if (debug_) {
+        LOG_DEBUG(absl::StrCat(
+            "Added [", token, "] ", attribute_gen_config.output_attribute(),
+            " if (", matchconfig.condition(), ") -> ", matchconfig.value()));
+      }
+
+      tokens_.push_back(token);
+      matches.push_back(
+          Match(matchconfig.condition(), token, matchconfig.value()));
+    }
+    gen_.push_back(AttributeGenerator(
+        phase, attribute_gen_config.output_attribute(), std::move(matches)));
+    matches.clear();
+  }
+  return true;
+}
+
+void PluginRootContext::cleanupAttributeGen() {
+  gen_.clear();
+  for (const auto& token : tokens_) {
+    exprDelete(token);
+  }
+  tokens_.clear();
+}
+
+bool PluginRootContext::onDone() {
+  cleanupAttributeGen();
+  return true;
+}
+
+// attributeGen is called on the data path.
+void PluginRootContext::attributeGen(EvalPhase phase) {
+  for (const auto& attribute_generator : gen_) {
+    if (phase != attribute_generator.phase()) {
+      continue;
+    }
+
+    std::string val;
+    auto eval_status = attribute_generator.evaluate(&val);
+    if (!eval_status) {
+      incrementMetric(runtime_errors_, 1);
+
+      // eval failed set error attribute
+      // __error attribute should be used by downstream plugins to distinguish
+      // between
+      // 1. No conditions matched.
+      // 2. Error evaluating a condition.
```

Comment:
Why does it matter to downstream plugins? Either way, there is no value so the outcome is the same.

Reply:
+1. Can you elaborate some cases that you think it would be useful?

Reply:
checking for errors is often easier than remembering to reason about absence of values, imho. any policy-like extensions relying on attribute-gen may only want to operate on a full set. is there another way to signal error in generation?